test/common/upstream/cluster_manager_impl_test.cc

```diff
@@ -1620,6 +1693,429 @@ TEST_F(ClusterManagerImplTest, OriginalDstInitialization) {
   factory_.tls_.shutdownThread();
 }
 
+TEST_F(ClusterManagerImplTest, MergedUpdates) {
+  createWithLocalClusterUpdate();
+
+  // Ensure we see the right set of added/removed hosts on every call.
+  EXPECT_CALL(local_cluster_update_, post(_, _, _))
+      .WillRepeatedly(Invoke([](uint32_t priority, const HostVector& hosts_added,
+                                const HostVector& hosts_removed) -> void {
+        static int call = 0;
+
+        switch (call) {
+        case 0:
+          EXPECT_EQ(0, priority);
+          EXPECT_EQ(0, hosts_added.size());
+          EXPECT_EQ(1, hosts_removed.size());
+          break;
+        case 1:
+          EXPECT_EQ(0, priority);
+          EXPECT_EQ(0, hosts_added.size());
+          EXPECT_EQ(1, hosts_removed.size());
+          break;
+        case 2:
+          EXPECT_EQ(0, priority);
+          EXPECT_EQ(2, hosts_added.size());
+          EXPECT_EQ(0, hosts_removed.size());
+          break;
+        }
+
+        ++call;
+      }));
+
+  const Cluster& cluster = cluster_manager_->clusters().begin()->second;
+  HostVectorSharedPtr hosts(
+      new HostVector(cluster.prioritySet().hostSetsPerPriority()[0]->hosts()));
+  HostsPerLocalitySharedPtr hosts_per_locality = std::make_shared<HostsPerLocalityImpl>();
+  HostVector hosts_added{};
+  HostVector hosts_removed{};
+
+  // No timer needs to be mocked at this point, since the first update should be applied
+  // immediately.
+  hosts_removed.push_back((*hosts)[0]);
+  cluster.prioritySet().hostSetsPerPriority()[0]->updateHosts(
+      hosts, hosts, hosts_per_locality, hosts_per_locality, {}, hosts_added, hosts_removed);
+
+  // Now we need the timer to be ready, since createTimer() will be call given that this update
+  // _will_ be delayed.
+  Event::MockTimer* timer = new NiceMock<Event::MockTimer>(&factory_.dispatcher_);
+  hosts_removed.clear();
+  hosts_removed.push_back((*hosts)[1]);
+  cluster.prioritySet().hostSetsPerPriority()[0]->updateHosts(
+      hosts, hosts, hosts_per_locality, hosts_per_locality, {}, hosts_added, hosts_removed);
+
+  // Ensure the merged updates were applied.
+  timer->callback_();
+  EXPECT_EQ(1, factory_.stats_.counter("cluster_manager.merged_updates").value());
+
+  // Prepare a new timer for the next round of updates.
+  timer = new NiceMock<Event::MockTimer>(&factory_.dispatcher_);
```

Comment:
Arguably, we could do the small optimization I mentioned earlier about reusing timers (have some flag to let you know if its set etc.).